feat: add dynamic CRD watcher using informer

dhimanAbhi opened the pull request:
## Explanation
This PR introduces improvements to Kyverno's CRD cache behavior by allowing users to enable a CRD watcher on discovery cache to invalidate the discovery cache whenever we Add, Update or Delete any CRD.
<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->

## Related issue
Closes #10729
<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
/milestone 1.15.0
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## Documentation (required for features)

My PR contains new or altered behavior to Kyverno. 
- [ ] I have sent the draft PR to add or update [the documentation](https://github.com/kyverno/website) and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->

## What type of PR is this
/kind feature
<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->

## Proposed Changes
A new CRD Watcher to watch over crd. It can be enbaled by ```--set admissionController.crdWatcher=true``` while installation via helm. The CRD watcher is built using the Kubernetes dynamic client and informer mechanism. This allows Kyverno to automatically invalidate its discovery cache whenever CRDs change, making it more responsive and always up to date.

To enable CRD watching, Kyverno needs the following RBAC permissions:
```
- apiGroups: ["apiextensions.k8s.io"]
  resources: ["customresourcedefinitions"]
  verbs: ["get", "list", "watch"]
```
Without these permissions, the CRD watcher will not be able to function properly.
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests
Following logs generated when I added new CRD and crd watcher immediabled handled AddFunc event to invalidated discovery cache.
```
$ kubectl logs -n kyverno kyverno-admission-controller-6bb9f775c5-sl9g9  -f | grep "CRD added"
Defaulted container "kyverno" out of: kyverno, kyverno-pre (init)
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=cleanuppolicies.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=clustercleanuppolicies.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=clusterephemeralreports.reports.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=clusterpolicies.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=clusterpolicyreports.wgpolicyk8s.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=ephemeralreports.reports.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=globalcontextentries.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=imagevalidatingpolicies.policies.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=policies.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=policyexceptions.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=policyexceptions.policies.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=policyreports.wgpolicyk8s.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=updaterequests.kyverno.io namespace= v=0
2025-05-14T13:51:16Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=validatingpolicies.policies.kyverno.io namespace= v=0
2025-05-14T13:56:45Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:115 > CRD added logger=dynamic-client/crd-definition-watcher name=foos.samplecontroller.k8s.io namespace= v=0
```

SImilarly, DeleteFunc event was triggered that invalidated discovery cache when CRD was deleted. 
```
$ kubectl logs -n kyverno kyverno-admission-controller-6bb9f775c5-sl9g9  -f | grep "CRD deleted"
Defaulted container "kyverno" out of: kyverno, kyverno-pre (init)
2025-05-14T13:57:45Z INF github.com/kyverno/kyverno/pkg/clients/dclient/discovery.go:135 > CRD deleted logger=dynamic-client/crd-definition-watcher name=foos.samplecontroller.k8s.io namespace= v=0

```

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [ ] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [ ] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

eddycharly commented on pkg/clients/dclient/client.go:
```diff
@@ -85,6 +87,12 @@ func NewClient(
+	// This watcher will watch for CRD changes and invalidate the cache on changes on sutomresourcedefinitions
+	if crdWatcher {
+		log.Println("Starting the CRD Watcher...")
+		go discoveryClient.CreateCRDWatcher(ctx, dyn)
```
We need proper error handling here

realshuting reviewed (changes requested):
Can we add required permissions in Helm Chart when the watcher is enabled?

And there are linter issues and code-gen files to fix.

dhimanAbhi commented:
Hi @realshuting @eddycharly,

I’ve made the requested changes and implemented the following fixes:

 - Used metadatainformer.NewSharedInformerFactory with *metav1.PartialObjectMetadata to ensure only metadata is fetched, reducing CRD payload size.

 - Implemented proper error handling for CreateCRDWatcher..

 - Resolved all linting issues and ensured generated code aligns with the changes.

Please let me know if any further changes are needed.
Thanks!

realshuting reviewed (approved):
/lgtm

Thanks @dhimanAbhi for your contribution!

realshuting commented:
/cherry-pick release-1.14
